keycloak/keycloak-benchmark
# Active/Active XSite fencing. Resolves keycloak#29303

Resolves keycloak/keycloak#29303

## Changes
- User alert routing enabled on ROSA clusters

- PrometheusRule used to trigger AWS Lambda webhook in the event of a
  split-brain so that only a single site remains in the global accelerator endpoints

- Global Accelerator scripts refactored to use OpenTofu when creating
  AWS resources

- Task created to deploy/undeploy Active/Active

- Task created to simulate split-brain scenarios

- 'active-active' flag added to GH actions to differentiate between
  active/passive and active/active deployments

## Global Accelerator Provisioning

The global accelerator provisioning uses a hybrid approach for creating AWS resources. The NLB required for the accelerator endpoints is created via Kubernetes LoadBalancer services in each of the nodes. This is done as it's *much* simpler than trying to explicitly provision NLBs for each sites using OpenTofu. Consequently, the OpenTofu accelerator module simply references these existing NLBs via data sources so that we can add them to the accelerator endpoint group.

## Testing
1. Provision an active/active deployment:

```
gh workflow run rosa-multi-az-cluster-create.yml -f activeActive=true -f clusterPrefix= -f region=
```

2. Inspect the AWS Global Accelerator console and ensure that the endpoint group contains two endpoints, one for each site.

3. Simulate a split-brain scenario:

```
cd provision/infinispan
PREFIX= ROSA_CLUSTER_NAME_1=$PREFIX-a ROSA_CLUSTER_NAME_2=$PREFIX-b NAMESPACE=runner-keycloak task crossdc-split
```

4. Navigate to the Openshift Console and ensure an event was fired, go to  Observer -> Alerting and apply the "user" filter. A "SiteOffline" event should have been fired

5.  Inspect the AWS Global Accelerator console and ensure that the endpoint group now only contains a single endpoint.

## TODO

Still missing:
- [x] Infinispan 15.0.4.Final: https://github.com/infinispan/infinispan/pull/12368
- [ ] Scaling Benchmark Integration
- [ ] Webhook authentication
- [ ] Infinispan 15.0.5.Final: https://github.com/infinispan/infinispan/pull/12484
